### DISCUSSION TITLE
CPM and Submodules Git

### DISCUSSION THREAD

**Discussion by Andres6936** [cpm-cmake/CPM.cmake#188] [Q&A]
Very good morning, the following is a couple of questions that have arisen from a structuring that my project has suffered.

How does the configuration of a project work when it has several or a single git submodule/s? Does CPM download these submodules? If the above question is yes, is there a way to prevent CPM from downloading these submodules?

To give a little more context, I happen to have divided my [project](https://github.com/Andres6936/Doryen) into submodules, but these submodules do not contain anything that can stop the compilation process and use of my project (library), because they do not contain code and only contains the documentation of the project, I have done so (using submodules), because I am currently using CPM to download and configure that dependency automatically, see the following projects.

- [Hippy Test](https://github.com/Andres6936/Hippy.Doryen)
- [The Boring Game](https://github.com/Andres6936/Cataclysm)

The issue is that, currently my project contains a website (where the documentation resides) and a couple of directories containing images (fonts), however, external projects (other people's projects) using my library will probably not need all those fonts at first, Another reason for using submodules is the possibility of reducing the time it takes for CPM (Fetch_Content) to download the project, hence the question of how to avoid (in case it really happens) that CPM downloads these submodules.

But there is also another reason that leads me to separate the project into submodules, such as the convenience of separating issues and Pull Request for each section, as well as the opinion I have about how my project should be used, that is, as a downloadable dependency from the Internet and not installable on the system (although no one prevents you from doing so).

**Comment by TheLartians** [comment]
As CPM.cmake forwards its arguments to CMake's [`FetchContent`](https://cmake.org/cmake/help/latest/module/FetchContent.html), which forwards them to [`ExternalProject`](https://cmake.org/cmake/help/latest/module/ExternalProject.html) the behaviour should be about the same as declaring the dependency there. From the [docs](https://cmake.org/cmake/help/latest/module/ExternalProject.html):

```
GIT_SUBMODULES <module>...
```

Specific git submodules that should also be updated. If this option is not provided, all git submodules will be updated. When `CMP0097` is set to `NEW` if this value is set to an empty string then no submodules are initialized or updated.

```
GIT_SUBMODULES_RECURSE <bool>
```

Specify whether git submodules (if any) should update recursively by passing the --recursive flag to git submodule update. If not specified, the default is on.

So it does appear that by default CPM.cmake should indeed initialise a dependencies submodules. As for disabling the recursive clone, it seems that setting `GIT_SUBMODULES` to an empty string might work, however I'm uncertain if argument forwarding will work properly. Alternatively, you could also try [setting it to an existing non-submodule directory](https://gitlab.kitware.com/cmake/cmake/-/issues/20579#note_734045).

**Reply by bjconlan** [reply]
On a related note, I noticed that using `GIT_SUBMODULES ""` isn't preserving the string (used as a way to prevent submodules from being resolved (https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4729). I could work around this specifying a patsy folder in the repo but would be nice to have the FetchContent behavior. (this was using cpm 0.40.8 and cmake 3.25.1)
